src/workload.py

```diff
@@ -223,6 +224,31 @@ def _router_username(self) -> str:
         )
         return config["metadata_cache:bootstrap"]["user"]
 
+    def _wait_until_mysql_router_ready(self) -> None:
+        """Wait until a connection to MySQL Router is possible.
+
+        Retry every 5 seconds for up to 30 seconds.
+        """
+        logger.debug("Waiting until MySQL Router is ready")
+        self._charm.set_status(
+            event=None, additional_statuses=[ops.WaitingStatus("MySQL Router starting")]
+        )
+        try:
+            for attempt in tenacity.Retrying(
+                reraise=True,
+                stop=tenacity.stop_after_delay(30),
+                wait=tenacity.wait_fixed(5),
+            ):
+                with attempt:
+                    for port in [6446, 6447]:
+                        with socket.socket() as s:
+                            assert s.connect_ex(("localhost", port)) == 0
```

Comment:
I personally prefer not to use `assert` outside of tests, you could use [`s.connect`](https://docs.python.org/3/library/socket.html#socket.socket.connect) and have an exception raised instead of testing on the return value. 
This way, your retry will be triggered the same way but because of an exception related to to the socket module.

Reply:
`assert` should be avoided outside of tests, you can read a [small summary](https://stackoverflow.com/a/1838411/9997227) about the topic, use `if` and raise an custom exception if needed.

---
> We don't want to catch all exceptions

Your Retrying logic already does it partly (within the set timeout), and in general - you should handle them. But in your specific case, I'm not sure I see the value in differentiating between raised socket related exceptions.

This is what I'm suggesting as a simpler alternative approach:

1. **helper_networking.py:** (because testing on the network has not much to do with the charm logic)
    ```
    def is_reachable(host: str, port: int) -> bool:
        s = socket.socket()
        s.settimeout(30)   # This replaces all your retry logic
        try:
            s.connect((host, port))
        except Exception as e:  # you can list all exceptions, though again, I don't see the value
            logger.error(e)
            return False
        finally:
            s.close()
    ```

Then in your charm:
```
if not (is_reachable("localhost", 9446) and is_reachable("localhost", 9447)):
    raise MySQLRouterConnectionError()
```